src/atomate2/aims/jobs/base.py

```diff
@@ -0,0 +1,313 @@
+
+        Parameters
+        ----------
+            atoms : .MSONableAtoms
+                a structure to run a job
+        """
+        convergence_job = self.convergence_iteration(atoms)
+        return Flow([convergence_job], output=convergence_job.output)
+
+    @job
+    def convergence_iteration(
+        self,
+        structure: MSONableAtoms | Structure | Molecule,
+        prev_dir: str | Path = None,
+    ) -> Response:
+        """Run several jobs with changing inputs to test convergence.
+
+        Parameters
+        ----------
+        structure : .MSONableAtoms or Structure or Molecule
+            The structure to run the job for
+        prev_dir: str or None
+            An FHI-aims calculation directory in which previous run contents are stored
+
+        Returns
+        -------
+        The output response for the job
+        """
+        if isinstance(structure, (Structure, Molecule)):
+            atoms = MSONableAtoms.from_pymatgen(structure)
+        else:
+            atoms = structure.copy()
+
+        # getting the calculation index
+        idx = 0
+        converged = False
+        if prev_dir is not None:
+            prev_dir_no_host = str(prev_dir).split(":")[-1]
+            convergence_file = Path(prev_dir_no_host) / CONVERGENCE_FILE_NAME
+            idx += 1
+            if convergence_file.exists():
+                with open(convergence_file) as f:
+                    data = json.load(f)
+                    idx = data["idx"] + 1
+                    # check for convergence
+                    converged = data["converged"]
+        else:
+            prev_dir_no_host = None
+        if idx < self.last_idx and not converged:
+            # finding next jobs
+            next_base_job = self.maker.make(atoms, prev_dir=prev_dir_no_host)
+            next_base_job.update_maker_kwargs(
+                {
+                    "_set": {
+                        f"input_set_generator->user_parameters->"
+                        f"{self.convergence_field}": self.convergence_steps[idx]
+                    }
+                },
+                dict_mod=True,
+            )
+            next_base_job.append_name(append_str=f" {idx}")
+
+            update_file_job = self.update_convergence_file(
+                prev_dir=prev_dir_no_host,
+                job_dir=next_base_job.output.dir_name,
+                output=next_base_job.output,
+            )
+
+            next_job = self.convergence_iteration(
+                atoms, prev_dir=next_base_job.output.dir_name
+            )
+
+            replace_flow = Flow(
+                [next_base_job, update_file_job, next_job], output=next_base_job.output
+            )
+            return Response(detour=replace_flow, output=replace_flow.output)
+        task_doc = AimsTaskDoc.from_directory(prev_dir_no_host)
+        return ConvergenceSummary.from_aims_calc_doc(task_doc)
+
+    @job(name="Writing a convergence file")
+    def update_convergence_file(
```

Comment:
Typically, `Makers` don't have more than one function that produces a job. Unless there is a very good reason to, I would refactor this out to just be a function rather than an instance method.

Reply:
But I need class attributes for the job. After refactoring I'll have to present them as arguments to the function, which will bloat the code and contradict the DRY principle.

Reply:
I don't think it will result in that much code bloat, it just means you will pass 2 more arguments to the function.

Reply:
I don't think it will result in bloat, it is simply that two more arguments will get added to the function.